# First time, a Real Datacleaning experience

In [1]:
import pandas as pd
import numpy as np

In [4]:
!curl -O https://s3.amazonaws.com/openrecipes/20170107-061401-recipeitems.json.gz              #Downloaded the File

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0     0    0     0    0     0      0      0 --:--:--  0:00:01 --:--:--     0
  0 29.3M    0 17040    0     0   7174      0  1:11:32  0:00:02  1:11:30  7174
  0 29.3M    0  101k    0     0  31780      0  0:16:09  0:00:03  0:16:06 31780
  0 29.3M    0  254k    0     0  61023      0  0:08:24  0:00:04  0:08:20 61023
  1 29.3M    1  458k    0     0  86427      0  0:05:56  0:00:05  0:05:51 91763
  2 29.3M    2  611k    0     0  99987      0  0:05:07  0:00:06  0:05:01  130k
  2 29.3M    2  764k    0     0   102k      0  0:04:53  0:00:07  0:04:46  147k
  2 29.3M    2  900k    0     0   106k      0  0:04:43  0:00:08  0:04:35  153k
  3 29.3M    3 1019k    0     0   110k      0  0:04

In [5]:
!gunzip 20170107-061401-recipeitems.json.gz                                              # Unzipped and renamed Recipe

### Now, let's explore!

In [9]:
pd.read_json('Recipe.json')

ValueError: Trailing data

### Oops! We've got an Error!

• The error came because we have json file consisting more than one `line` and each line is ending with `\n` instead of `','`.<br>
• To resolve that, we have the **Easy way** ( which has been added in the new Pandas `add lines=True parameter`.
<br>
• **Another way** is to add `','` manually at the end of each lines.

More discussions at: [Stackoverflow](https://stackoverflow.com/questions/30088006/loading-a-file-with-more-than-one-line-of-json-into-pandas)

#### Easy Way

In [11]:
pd.read_json('Recipe.json', lines= True).head(2)

_id                             name  \
0  {'$oid': '5160756b96cc62079cc2db15'}  Drop Biscuits and Sausage Gravy   
1  {'$oid': '5160756d96cc62079cc2db16'}        Hot Roast Beef Sandwiches   

                                         ingredients  \
0  Biscuits\n3 cups All-purpose Flour\n2 Tablespo...   
1  12 whole Dinner Rolls Or Small Sandwich Buns (...   

                                                 url  \
0  http://thepioneerwoman.com/cooking/2013/03/dro...   
1  http://thepioneerwoman.com/cooking/2013/03/hot...   

                                               image  \
0  http://static.thepioneerwoman.com/cooking/file...   
1  http://static.thepioneerwoman.com/cooking/file...   

                         ts cookTime           source recipeYield  \
0  {'$date': 1365276011104}    PT30M  thepioneerwoman          12   
1  {'$date': 1365276013902}    PT20M  thepioneerwoman          12   

  datePublished prepTime                                        description  \
0    2013-03-11    PT10M  Late Saturday afternoon, after Marlboro Man ha...   
1    2013-03-13    PT20M  When I was growing up, I participated in my Ep...   

  totalTime creator recipeCategory dateModified recipeInstructions  
0       NaN     NaN            NaN          NaN                NaN  
1       NaN     NaN            NaN          NaN                NaN

### Another Way

In [70]:
with open('Recipe.json',  encoding="utf8") as json:
    data = (line.strip() for line in json)    #Makes a Generator so it doesn't occupy the space
    data_json = []
    for line in data:
        data_json.append(eval(line))

NameError: name 'null' is not defined

    Both ↑ and ↓ ways are good, but there is the 'null' values, which are becoming a bullshit.

In [50]:
with open('Recipe.json',  encoding="utf8") as f:
    lines = f.readlines()
    
# Now we have [ '{}', '{}', ... '{}' ]
lines = [eval(line) for line in lines]
pd.DataFrame(lines)

NameError: name 'null' is not defined

###  So, from the Book!

In [94]:
with open('Recipe.json', 'r', encoding= 'U8') as json:
    data = (line.strip() for line in json)
    data_json = "{0}".format(','.join(data)) # ',' join on Iterable! That is generator!
    
len(data_json)                                 # Total characters! As json file has to be in TOTAL STRING!

141105255

In [99]:
pd.read_json(data_json)

ValueError: Protocol not known:  and do a little sexy twirl?  Such a flirty little thing!\n\n\"Asian Lettuce Wraps\" or \"Asian Lettuce Cups\" is the most requested recipe on my site...." },{ "_id" : { "$oid" : "5428a17596cc6211615607e2" }, "name" : "Baby Bok Choy with Ground Chicken Stir Fry", "ingredients" : "2 teaspoons cooking oil\n1/2 pound ground chicken\n1 stalk green onion, chopped\n2 teaspoons grated fresh ginger\n2 cloves garlic, finely minced\n1 pound bok choy, cut in half lengthwise\n1 tablespoon water\n2 teaspoons soy sauce\n1 teaspoon sesame oil", "url" : "http

# So, yeah! That didn't work as well.
In new pandas, they have changed the `read_json` functionality and that's why it gives an error. More: [Here](https://stackoverflow.com/questions/63553845/pandas-read-json-valueerror-protocol-not-known)

**In short: We will use its built in `lines = True` just fine!**

###  

###  

In [2]:
#Yes, that easy way...
recipes = pd.read_json('Recipe.json', lines= True)

In [3]:
recipes.shape

(173278, 17)

In [5]:
recipes.iloc[1]

_id                                {'$oid': '5160756d96cc62079cc2db16'}
name                                          Hot Roast Beef Sandwiches
ingredients           12 whole Dinner Rolls Or Small Sandwich Buns (...
url                   http://thepioneerwoman.com/cooking/2013/03/hot...
image                 http://static.thepioneerwoman.com/cooking/file...
ts                                             {'$date': 1365276013902}
cookTime                                                          PT20M
source                                                  thepioneerwoman
recipeYield                                                          12
datePublished                                                2013-03-13
prepTime                                                          PT20M
description           When I was growing up, I participated in my Ep...
totalTime                                                           NaN
creator                                                         

### Good way... Remember it ↓

In [7]:
recipes['ingredients'].str.len().describe()

count    173278.000000
mean        244.617926
std         146.705285
min           0.000000
25%         147.000000
50%         221.000000
75%         314.000000
max        9067.000000
Name: ingredients, dtype: float64

The maximumn number of Ingredients are 9067... let's find which one is that!

In [12]:
np.max(recipes['ingredients'].str.len())

9067

In [13]:
np.argmax(recipes['ingredients'].str.len())

135598

#### <font face='nexa'> I LIKE THIS LINE!

In [10]:
recipes['name'][ np.argmax( recipes['ingredients'].str.len() ) ]

'Carrot Pineapple Spice &amp; Brownie Layer Cake with Whipped Cream &amp; Cream Cheese Frosting and Marzipan Carrots'

# 

See which are *Breakfast* foods `str.contains`

In [17]:
recipes['description'].str.contains("[Bb]reakfast").sum()        # See the [Bb] here!!!!

3524

# 

# 

### <font face='nexa'> NOW LET'S HAVE `A little` HICCUP!
Building a recomendation system!

How it works?
    
    This is bases on the Ingredients that we put in, so if the recipe contains those,
    it will give a reply!
    
    Simple!

In [110]:
recipes.loc[[34,456,334], 'ingredients']

34     1 medium sweet potato\n½ medium onion\n⅓ cup g...
456    2 Marcona almonds\n1 olive oil\n1 smoked papri...
334    4 large potatoes, preferably Maris Piper \n, f...
Name: ingredients, dtype: object

In [32]:
tempDF = pd.DataFrame({'ings': recipes['ingredients'].str.split('\n') })

In [35]:
from collections import Counter

In [50]:
temp = list(tempDF['ings'].iloc[0:10])

In [56]:
final = []
for i in temp:
    for kaka in i:
        for j in kaka.split():
            final.append(j.strip())

In [58]:
Counter(final)

Counter({'Biscuits': 2,
         '3': 4,
         'cups': 6,
         'All-purpose': 3,
         'Flour': 3,
         '2': 13,
         'Tablespoons': 6,
         'Baking': 2,
         'Powder': 2,
         '1/2': 7,
         'teaspoon': 11,
         'Salt': 4,
         '1-1/2': 2,
         'stick': 2,
         '(3/4': 2,
         'Cup)': 2,
         'Cold': 3,
         'Butter,': 2,
         'Cut': 2,
         'Into': 2,
         'Pieces': 2,
         '1-1/4': 2,
         'cup': 19,
         'Butermilk': 1,
         'SAUSAGE': 1,
         'GRAVY': 1,
         '1': 16,
         'pound': 3,
         'Breakfast': 1,
         'Sausage,': 1,
         'Hot': 2,
         'Or': 4,
         'Mild': 1,
         '1/3': 4,
         '4': 3,
         'Whole': 2,
         'Milk': 1,
         'Seasoned': 1,
         'teaspoons': 2,
         'Black': 1,
         'Pepper,': 3,
         'More': 1,
         'To': 2,
         'Taste': 1,
         '12': 1,
         'whole': 5,
         'Dinner': 1,
       

Here I tried to do like this, I mean to get the Ingredients, one by one and ↑ just 10 records! So, not feasible.
Need Machine Learning for this...

Following the book's way...

In [59]:
ingredients = ['salt', 'pepper', 'papper', 'oregano', 'sage', 'rosemary', 'tarragon', 'paprika']

In [95]:
availablein = dict( ((i, recipes['ingredients'].str.contains( i )) for i in ingredients))

In [96]:
availablein = pd.DataFrame(availablein)
availablein

salt  pepper  papper  oregano   sage  rosemary  tarragon  paprika
0       False   False   False    False   True     False     False    False
1       False   False   False    False  False     False     False    False
2        True    True   False    False  False     False     False    False
3       False   False   False    False  False     False     False    False
4       False   False   False    False  False     False     False    False
...       ...     ...     ...      ...    ...       ...       ...      ...
173273  False   False   False    False  False     False     False    False
173274  False   False   False    False  False     False     False    False
173275   True    True   False    False  False     False     False     True
173276   True    True   False    False  False     False     False    False
173277   True   False   False    False  False     False     False    False

[173278 rows x 8 columns]

In [103]:
# Now, getting the requirements...
availablein[['rosemary', 'paprika', 'oregano']].sum()

rosemary    3039
paprika     3528
oregano     3468
dtype: int64

No, that shows where the `rosemary, paprika, oreganor` are available! <br>
We want only whose where these three are together!

In [106]:
selection = availablein[ (availablein['rosemary']) & (availablein['paprika']) & (availablein['oregano'])]

In [107]:
len(selection)

17

Getting their names...

In [114]:
recipes.name[ selection.index ]

66245                                    Emeril's Beef Stew
69307                         Vegetarian Five Bean Soup Mix
70209     Tapas feast: tortilla, glazed chorizo, mancheg...
71443     Tapas feast: tortilla, glazed chorizo, mancheg...
76489                                   Pasta alla Giovanni
86870                                           Magic Sauce
90652                    Balsamic Marinated Chicken Breasts
92889         Book Club Herb Roasted Chicken and Vegetables
95737                              Buttermilk Fried Chicken
105548                         Garlic Herb Skillet Potatoes
110866                                           Kas' Chili
112590      Mama's Oh-So-Savory Lamb and Eggplant Casserole
118282                      Peruvian Style Beer Can Chicken
121140                           Red Pepper and Tomato Soup
121811    Roasted Pork Chops with Tomatoes, Mushrooms, a...
125461                                 Special Italian Meat
150598                        Vegetarian

# 

# <font face='consolas' size=6> Yeah I know that this is not entierly done by me, I have referenced the book, but, Still I learnt a lot, and this is my first data cleaning task! 
  